sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -528,43 +528,52 @@ pi_result cuda_piPlatformsGet(pi_uint32 num_entries, pi_platform *platforms,
                               pi_uint32 *num_platforms) {
 
   try {
-    static constexpr pi_uint32 numPlatforms = 1;
+    static std::once_flag initFlag;
+    static pi_uint32 numPlatforms = 1;
+    static _pi_platform platformId;
 
-    if (num_platforms != nullptr) {
-      *num_platforms = numPlatforms;
+    if (num_entries == 0 and platforms != nullptr) {
+      return PI_INVALID_VALUE;
+    }
+    if (platforms == nullptr and num_platforms == nullptr) {
+      return PI_INVALID_VALUE;
     }
 
     pi_result err = PI_SUCCESS;
 
-    if (platforms != nullptr) {
-
-      assert(num_entries != 0);
-
-      static std::once_flag initFlag;
-      static _pi_platform platformId;
-      std::call_once(
-          initFlag,
-          [](pi_result &err) {
-            err = PI_CHECK_ERROR(cuInit(0));
-
-            int numDevices = 0;
-            err = PI_CHECK_ERROR(cuDeviceGetCount(&numDevices));
-            platformId.devices_.reserve(numDevices);
-            try {
-              for (int i = 0; i < numDevices; ++i) {
-                CUdevice device;
-                err = PI_CHECK_ERROR(cuDeviceGet(&device, i));
-                platformId.devices_.emplace_back(
-                    new _pi_device{device, &platformId});
-              }
-            } catch (...) {
-              // Clear and rethrow to allow retry
-              platformId.devices_.clear();
-              throw;
+    std::call_once(
+        initFlag,
+        [](pi_result &err) {
+          CUresult result = cuInit(0);
+          if (result == CUDA_ERROR_NO_DEVICE) {
```

Comment:
it's not in the docs, but it is returned when there are no available devices (I'm testing with CUDA 10.2).

Reply:
The idea was to special case a working installation which simply has no devices (for example because of `CUDA_VISIBLE_DEVICES=`).

I guess it also makes sense not to return the platform, instead of calling`abort()`, if `cuInit()` fails for other reasons.

I can make the change (and remove the call to `PI_CHECK_ERROR(result)`) if this is the preferred behaviour.

Reply:
I would argue that `cuInit()` should be allowed to fail for **any reason**. At that point we should just not get any CUDA devices at runtime. The application should still be allowed to sue OpenCL/host devices at that point.

So yeah, I would just check with `CUDA_SUCCESS`.

Reply:
I can confirm CUDA_ERROR_NO_DEVICE is reported by cuInit, despite of it not being documented. 
I think some of the other error causes are worth of throwing error, otherwise users will simply get "no cuda platform" without any other information.

Reply:
:confused: Do you think production code should behave that way? If I want to run an application on "some platform", I mainly want it to run. If it can't use any CUDA devices, then fine, it should still use any other device that it can, and not just die on me.

So an exception is really not what I would go for here. For whatever reason the `cuInit()` function is dying. At the most I would print an "info message" if the return value of `cuInit()` is something **really** non-expected.

Reply:
> I think some of the other error causes are worth of throwing error, otherwise users will simply get "no cuda platform" without any other information.

I do not know how the plugins should work in the long term.
For the moment I understand that - if they are built - both the OpenCL and the CUDA plugins will be used, without a way to select/deselect them at runtime.

Now, the OpenCL plugin seems happy in not reporting any platforms for which it doesn't find any devices.

I think the CUDA plugin should never abort (or throw an exception that causes the runtime to do so); for example because that would break also SYCL applications that do not care about the PTX/CUDA backend.

Whether to print an error or silently report no CUDA devices (and thus no CUDA platform) is less clear, but I would suggest to use the same approach as the OpenCL plugin (i.e. be silent) and let the application or the user deal with the lack of devices.

If, on the other hand, one prefers a different behaviour (e.g. warning about a CUDA installation without device, or an NVIDIA device without CUDA drivers, etc.) I think the OpenCL backend should behave in the same way with respect to OpenCL devices.